EducationalTestingService/rsmtool
# Add type hints (Part 3)

### Changes

- Add `statsmodels` intersphinx mapping for documentation.
- add type hints to `fairness_utils.py`.
- add type hints to `modeler.py`.
- add type hints to `preprocessor.py`.
- add type hints to `reader.py`
- Remove incorrect exceptions in some tests based on changes in `reader.py`. 

Note that in this MR, I also did some refactoring:
- Removed unused `id_column` argument in `FeaturePreprocessor.filter_on_column()`. 
- Changed the behavior of `DataReader.locate_files()` to always return list of string instead of either a string or a list of strings + `None`s which was confusing and also caused issues when adding type hints. 
- I also had to change a couple of tests to check for the raising of the correct exception type based on the changes in `DataReader`.

### How to review
- Please check that I have not left any parameters/arguments/return values without type hints.
- Make sure that I didn't screw up anything when doing the refactoring. 
- Here's the [RTD build for this branch](https://rsmtool.readthedocs.io/en/326-type-hints-3). Check the "API Documentation" section (only for the files affected by this MR) to confirm that all types are linked to relevant pages from external documentations (e.g., Python, Numpy, SKLL, stats models etc.)
- Check for typos and/or weird issues in docstrings.
